File: 27 February 2025 (lab 5)/Activity selection problem.cpp
#include <iostream>
using namespace std;

struct Activity {
    int start, end;
};

void selectionSort(Activity arr[], int n)
{
    for (int i = 0; i < n - 1; i++)
    {
        int minIdx = i;
        for (int j = i + 1; j < n; j++)
        {
            if (arr[j].end < arr[minIdx].end)
            {
                minIdx = j;
            }
        }
        swap(arr[i], arr[minIdx]);
    }
}

void activitySelection(Activity arr[], int n)
{
    selectionSort(arr, n);
    cout << "Selected activities (start, end):\n";
    int lastEndTime = arr[0].end;
    cout << "(" << arr[0].start << ", " << arr[0].end << ")\n";

    for (int i = 1; i < n; i++)
    {
        if (arr[i].start >= lastEndTime)
        {
            cout << "(" << arr[i].start << ", " << arr[i].end << ")\n";
            lastEndTime = arr[i].end;
        }
    }
}

int main() {
    int n;
    cout << "Enter number of activities: ";
    cin >> n;
    Activity arr[n];
    cout << "Enter start and end times of activities:\n";
    for (int i = 0; i < n; i++)
    {
        cin >> arr[i].start >> arr[i].end;
    }
    activitySelection(arr, n);
    return 0;
}

File: Lab 5 (27 February 2025)/Activity selection problem.py
import time
import random
import matplotlib.pyplot as p

def activity_selection(start, end, n):
    activities = sorted(zip(start, end), key=lambda x: x[1])  # Sort by finish time
    selected = [activities[0]]
    last_end = activities[0][1]

    for i in range(1, n):
        if activities[i][0] >= last_end:
            selected.append(activities[i])
            last_end = activities[i][1]
    return selected

sizes = [10, 100, 1000, 2000, 5000, 10000]
times = []

for n in sizes:
    start_times = [random.randint(0, 1000) for _ in range(n)]
    end_times = [s + random.randint(1, 100) for s in start_times]

    start_time = time.time()
    activity_selection(start_times, end_times, n)
    end_time = time.time()

    times.append(end_time - start_time)

import math
k = times[0] / (sizes[0] * math.log2(sizes[0]))
onlogn = [k * n * math.log2(n) for n in sizes]

p.figure(figsize=(10, 5))
p.plot(sizes, times, marker="o", label="Measured Time", color="red")
p.plot(sizes, onlogn, linestyle="--", label="O(n log n) Reference", color="blue")

p.xlabel("Number of Activities (n)")
p.ylabel("Time (seconds)")
p.title("Activity Selection: Time Complexity vs O(n log n)")
p.legend()
p.grid(True)
p.show()